### PR TITLE
feat(#8909): add CI jobs that run integration tests over a CHT launched in a K3D cluster 

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Adds two CI jobs to run integration tests over K3D architecture, using specific helm charts and local-path for persistent storage. 

#8909

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8909-k8s-e2e/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8909-k8s-e2e/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8909-k8s-e2e/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
Hi @nydr . It would be great to get some feedback on this, if you have the bandwidth. Thanks a lot!

**Review by nydr — APPROVED** [review]
This is so cool! I was able to run the suite locally without much effort and I see a lot of potential here 😃 

Am I correct in assuming a 30 minute extra build step is acceptable from a cost perspective?

I'm a bit concerned with the amount of potential false failures with flaky tests, if it becomes a big issue it might be worth considering skipping those by default in this suite.

Mostly minor comments, approving to unblock as I think it's good to merge as is. Overall impressive work!

File: tests/utils/index.js
Comment (dianabarsan) [reply tests/utils/index.js:1412]: What is the Kubernetes API? :D

Comment (nydr) [reply tests/utils/index.js:1412]: It's a rest API that kubectl (and other clients) is using behind the scenes to talk with Kubernetes. Easiest way is probably to use the official [kubernetes-client/javascript](https://github.com/kubernetes-client/javascript) which is being used in [upgrade-service-kubernetes (k8s-manager.ts)](https://github.com/medic/upgrade-service-kubernetes/blob/master/src/lib/k8s-manager.ts)

As a sidenote it is possible to run authorized API calls from inside the cluster by making use of [Kubernetes Service Account](https://kubernetes.io/docs/concepts/security/service-accounts/) (which needs a RoleBinding to a Role with the appropriate access). This is how the upgrade service manages its authorization when run in a cluster

That being said I think it's fine to offload the logic in tests to kubectl if it works reasonably well

Comment (dianabarsan) [reply tests/utils/index.js:1412]: I think I prefer kubectl logic in this case. I don't want to overcomplicate the CI suite.

File: .github/workflows/build.yml
Comment (dianabarsan) [reply .github/workflows/build.yml:125]: Yea, I'm not sure what to do here. 
I mainly wanted a suite that runs over k3d. I thought the integration suite (api + sentinel) is a good choice, mainly because I wanted to flush out most of the potential bottlenecks (scaling containers, following logs and etc), as these suites cover most of the serverside complexity. 
This is also in regards to your comment about whether we're fine with an 20+ minute build. The answer is: maybe not? Maybe it's just pointless to run integration over 2 architectures? Should we just change both integration suites to run over k3d? 
I think that would be my preferred option.

File: .github/workflows/build.yml
Comment (dianabarsan) [reply .github/workflows/build.yml:160]: I think our previous attempts of reusing code was unsuccessful, simply because GH code reuse is pretty cumbersome (like: https://docs.github.com/en/actions/creating-actions/creating-a-composite-action). 
I would love if this were just easily possible.

File: tests/utils/index.js
Comment (dianabarsan) [reply tests/utils/index.js:1167]: I thought about that too, but the code is so similar I thought it was a big shame to have all that overhead duplicated (spawn, hook 4 listeners that do the same thing, etc), so I opted for this instead.

File: tests/utils/index.js
Comment (dianabarsan) [reply tests/utils/index.js:1169]: for me it's more readable, rather than having a params array that splits everything by the space.

File: tests/utils/index.js
Comment (dianabarsan) [reply tests/utils/index.js:100]: I believe this is some leftover change from experiments around getting the correct docker gateway for k3d. Not a necessary change, but in itself inoffensive.

File: tests/integration/couchdb/couch_chttpd.spec.js
Comment (dianabarsan) [reply tests/integration/couchdb/couch_chttpd.spec.js:63]: I love this!

File: tests/utils/index.js
Comment (nydr) [reply tests/utils/index.js:1167]: Perfectly reasonable

File: tests/utils/index.js
Comment (nydr) [reply tests/utils/index.js:1412]: Fair enough

File: tests/utils/index.js
Comment (nydr) [reply tests/utils/index.js:1169]: Ok, it works for me as is

File: .github/workflows/build.yml
Comment (nydr) [reply .github/workflows/build.yml:125]: That works for me, I can't think of a reason integration tests need to be ran outside of k3d but a second opinion might be good

**Comment by dianabarsan** [comment]
Hi @nydr 
I addressed all comments and merged the templates into logical files. 
In terms of context (re conversation the other day), it seems that `k3d cluster create` automatically switches default context to the new cluster. While I don't think it's an issue in CI, I'm wondering if we should try avoiding that for local runs, where the user might want to return to their previous context. Any ideas?

**Comment by nydr** [comment]
> if we should try avoiding that for local runs, where the user might want to return to their previous context. Any ideas?

That would make sense, looks like it should be something like `k3c cluster create <cluster> -kubeconfig-update-default=false` and `--kubeconfig-switch-context` might need to be set to false too

File: tests/integration/.mocharc-k3d.js
Comment (dianabarsan) [reply tests/integration/.mocharc-k3d.js]: Added to the only testing readme that we have.

**Comment by dianabarsan** [comment]
I pushed a bunch of changes trying to stabilize and unify how the two systems work together. @nydr would you mind having a look?

**Review by nydr — APPROVED** [review]
LGTM assuming the test failures are due to flaky tests

As a sidenote, would it make sense to tag all flaky tests and run those as a separate step to speed up retries and highlight the tests with issues? I think they're quite disruptive but out of scope for this PR

**Comment by dianabarsan** [comment]
> LGTM assuming the test failures are due to flaky tests

I don't believe they are just flakes. I'm fixing them. It's been a journey....

File: tests/utils/index.js
Comment (dianabarsan) [reply tests/utils/index.js:1271]: I don't like this either. 
The k3d container takes a longer time to start - compared to docker - and logs are cleared when the container starts. If I start watching logs before the container is up, in k3d I get an error. If I add a wait and retry, and the operation I'm waiting for happens immediately after container startup (this is the case), I could start watching the logs after the action had happened (this is also the case). 

I'll add more info for the parameter.

**Pull request merged**
